domain/access/state/permission.go

```diff
@@ -195,74 +198,167 @@ func (st *PermissionState) UpsertPermission(ctx context.Context, args access.Upd
 
 // ReadUserAccessForTarget returns the subject's (user) access for the
 // given user on the given target.
+// accesserrors.PermissionNotFound is returned the users permission cannot be
+// found on the target.
 func (st *PermissionState) ReadUserAccessForTarget(ctx context.Context, subject string, target corepermission.ID) (corepermission.UserAccess, error) {
-	var userAccess corepermission.UserAccess
 	db, err := st.DB()
 	if err != nil {
-		return userAccess, errors.Trace(err)
+		return corepermission.UserAccess{}, errors.Trace(err)
+	}
+
+	user := dbPermissionUser{
+		Name: subject,
+	}
+	perm := dbPermission{
+		GrantOn: target.Key,
 	}
 
 	readQuery := `
-SELECT  (p.uuid, p.grant_on, p.grant_to, p.access_type, p.object_type) AS (&dbPermission.*),
-        (u.uuid, u.name, u.display_name, u.created_at, u.disabled) AS (&dbPermissionUser.*),
-        creator.name AS &dbPermissionUser.created_by_name
+SELECT  (u.uuid, u.name, u.display_name, u.external, u.created_at, u.disabled) AS (&dbPermissionUser.*),
+        creator.name AS &dbPermissionUser.created_by_name,
+        (p.*) AS (&dbPermission.*)
 FROM    v_user_auth u
         JOIN user AS creator ON u.created_by_uuid = creator.uuid
-        JOIN v_permission p ON u.uuid = p.grant_to
-WHERE   u.name = $permInOut.name
+        LEFT JOIN v_permission p ON u.uuid = p.grant_to AND p.grant_on = $dbPermission.grant_on 
+WHERE   u.name = $dbPermissionUser.name
 AND     u.disabled = false
 AND     u.removed = false
-AND     p.grant_on = $permInOut.grant_on
 `
 
-	readStmt, err := st.Prepare(readQuery, dbPermission{}, dbPermissionUser{}, permInOut{})
+	readStmt, err := st.Prepare(readQuery, user, perm)
 	if err != nil {
 		return corepermission.UserAccess{}, errors.Trace(err)
 	}
 
-	var (
-		readUser dbPermission
-		permUser dbPermissionUser
-	)
-
+	var baseExternalPerms dbPermission
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		in := permInOut{
-			Name:    subject,
-			GrantOn: target.Key,
-		}
-		err = tx.Query(ctx, readStmt, in).Get(&readUser, &permUser)
+
+		err = tx.Query(ctx, readStmt, user, perm).Get(&perm, &user)
 		if errors.Is(err, sql.ErrNoRows) {
-			return fmt.Errorf("%w for %q on %q", accesserrors.PermissionNotFound, subject, target.Key)
+			return errors.Annotatef(accesserrors.UserNotFound, "looking for permissions for %q on %q", subject, target.Key)
 		} else if err != nil {
 			return errors.Annotatef(err, "getting permission for %q on %q", subject, target.Key)
 		}
 
+		if user.External {
+			baseExternalPerms, err = st.baseExternalAccessForTarget(ctx, tx, target)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
 		return nil
 	})
 	if err != nil {
-		return userAccess, errors.Trace(err)
+		return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
```

Comment:
You no longer need to `CoerceError` in any of the state types. This is done automatically now.

Reply:
Oh sweet, which PR was that added in, I want to see how/where thats done.

Reply:
See: https://github.com/juju/juju/pull/17757